text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,1131 @@
+Resource types must also align: for example, a resource in V1 template of type `AWS::EC2::VPC` can only map to
+a resource in V2 template of a compatible type (e.g., `AWS::EC2::VPCV2`).
+
+This validation is specific to the context of construct upgrade framework and will be implemented as part of
+`cdk construct-upgrade --unstable=construct-upgrade` CLI command.
+
+> Note that CDK team is working on a [cdk refactor feature](https://github.com/aws/aws-cdk-rfcs/pull/705)
+> to support CloudFormation Stack refactoring with CDK CLI. Once implemented, we will onboard to use CDK
+> refactor feature instead and users do not need to manually create the `refactor.json` file and will be
+> inferred based on the construct tree and paths. This step is still needed to validate the inferred
+> refactor.json file by CDK.
+
+#### [In-Place Migration Only] Path Metadata Validation
+
+In In-Place migrations, we leverage the CloudFormation stack refactoring feature, which imposes a limitation:
+no changes to resource configurations are allowed, including updates to the Metadata section of any resource.
+
+A common issue arises when users migrate a construct across modules (i.e. `@aws-cdk.aws-ec2-alpha.VpcV2` vs
+`aws-cdk-lib.aws-ec2.Vpc`). Even if the underlying resources remain the same, construct path differences result
+in different `aws:cdk:path` metadata values in the output CloudFormation template.
+
+For In-Place migrations, because we're using CloudFormation stack refactoring feature, it does not allow any
+resource configuration changes, including `Metadata` in a resource definition. When users use V2 construct
+from a different module , the path metadata
+differs due to construct path differences.
+
+```yaml
+# VPC V1 - aws-cdk-lib.aws-ec2.Vpc
+vpcpublicSubnetSubnet1RouteTable155F8E2E:
+  Type: AWS::EC2::RouteTable
+  Properties:
+    ...
+  Metadata:
+    aws:cdk:path: CdkMigrationStack/vpc/publicSubnetSubnet1/RouteTable
+
+# VPC V2 - @aws-cdk/aws-ec2-alpha.VpcV2
+vpcpublicSubnetSubnet1RouteTable155F8E2E:
+  Type: AWS::EC2::RouteTable
+  Properties:
+    ...
+  Metadata:
+    aws:cdk:path: CdkMigrationStack/publicSubnet/RouteTable/RouteTable
+```
+
+This metadata change alone would cause the In-Place migration to fail, since CloudFormation treats it as
+a configuration change.
+
+To streamline this experience, the `cdk construct-upgrade` CLI will detect such cases. If path metadata
+is present in the output template and the chosen migration strategy is `In-Place migration`, the CLI will
+prompt the user:
+
+```sh
+Detected path metadata in the synthesized CloudFormation template, which is not compatible with CloudFormation
+stack refactoring. CDK will deploy your stack with path metadata disabled using: cdk deploy --no-path-metadata
+
+Would you like to proceed? (y/n)
+```
+
+If the user confirms by selecting "y", the tool will proceed to run `cdk deploy` with `--no-path-metadata` flag
```

Comment:
is `--no-path-metadata` an existing capability we have? will it just remove path metadata forever until a user deploys again without this flag?

Reply:
Yes it's an existing behaviour. I think this is the same type of behaviour as if we deploy the changes by setting RETAIN policy.